libcaf/libcaf/repository.py

```diff
@@ -691,6 +701,231 @@ def create_tag(self, tag_name: str, commit_hash: str, author: str, message: str)
     @requires_repo
     def status(self) -> Sequence[Diff]:
         return self.diff(self.head_commit(), self.working_dir)
+    
+    def _resolve_checkout_target(self, target: str) -> Ref | None:
+        """ Resolve a checkout target to a Ref.
+
+            The resolution is done by scanning the refs directory so we can distinguish between
+            branches and tags by location.
+            If there is a collision between a branch and a tag name, the branch takes precedence similair to Git.
+
+            :param target: The checkout target provided by the user.
+            :return: The resolved Ref object or None if the target does not exist.
+        """
+
+        if not isinstance(target, str):
+            msg = f'Invalid checkout target type: {type(target)}'
+            raise RefError(msg)
+
+        if len(target) == HASH_LENGTH and all(c in HASH_CHARSET for c in target):
+            return HashRef(target)
+
+        explicit_path = self.refs_dir() / target
+        if explicit_path.exists() and explicit_path.is_file():
+            if target.startswith(f'{TAGS_DIR}/'):
+                return TagRef(target)
+            return SymRef(target)
+
+        branch_path = self.heads_dir() / target
+        if branch_path.exists():
+            return branch_ref(target)
+
+        tag_path = self.tags_dir() / target
+        if tag_path.exists():
+            return tag_ref(target)
+
+        return None
+    
+    def _relpath_for(self, node: Diff) -> Path:
+        parts: list[str] = []
+        current: Diff | None = node
+        while current is not None and current.parent is not None:
+            if current.record.name:
+                parts.append(current.record.name)
+            current = current.parent
+        return Path(*reversed(parts))
+
+    def _read_object_bytes(self, hash_value: str) -> bytes:
+        with open_content_for_reading(self.objects_dir(), hash_value) as f:
+            return f.read()
+
+    def _record_at_path(self, root: Tree, rel_path: Path, tree_cache: dict[str, Tree]) -> TreeRecord:
+        current_tree = root
+        parts = list(rel_path.parts)
```

Comment:
You can derive the path from the tree itself, you don't need rel_path. This can be simplified.